File: CodeRun_Boost/7502_prob_win/solution.c
#include <stdio.h>

long long power(long long base, long long exp) {
    long long res = 1;
    long long MOD = 1e9 + 7;
    base %= MOD;
    while (exp > 0) {
        if (exp % 2 == 1) {
            res = (res * base) % MOD;
        }
        base = (base * base) % MOD;
        exp /= 2;
    }
    return res;
}

long long mod_inverse(long long n) {
    long long MOD = 1e9 + 7;
    return power(n, MOD - 2);
}

long long solve(int a, int k, int n) {
    long long MOD = 1e9 + 7;

    if (k == 0 || n == 0) {
        return 1;
    }

    long long small_val = (k < n) ? k : n;
    long long large_val = (k > n) ? k : n;

    long long numerator = 1;
    for (long long i = 1; i <= small_val; ++i) {
        numerator = (numerator * i) % MOD;
    }

    long long denominator = 1;
    for (long long i = 1; i <= small_val; ++i) {
        denominator = (denominator * (large_val + i)) % MOD;
    }
    
    return (numerator * mod_inverse(denominator)) % MOD;
}

File: CodeRun_Boost/7502_prob_win/solution.py
def solve(a: int, k: int, n: int) -> int:
    MOD = 10**9 + 7

    def power(base, exp):
        res = 1
        base %= MOD
        while exp > 0:
            if exp % 2 == 1:
                res = (res * base) % MOD
            base = (base * base) % MOD
            exp //= 2
        return res

    def mod_inverse(n):
        return power(n, MOD - 2)

    small_val = min(k, n)
    large_val = max(k, n)

    if small_val == 0:
        return 1

    numerator = 1
    for i in range(1, small_val + 1):
        numerator = (numerator * i) % MOD

    denominator = 1
    for i in range(1, small_val + 1):
        denominator = (denominator * (large_val + i)) % MOD
    
    return (numerator * mod_inverse(denominator)) % MOD